src/allocators/AllocationScope.zig

```diff
@@ -86,20 +111,33 @@ fn alloc(ctx: *anyopaque, len: usize, alignment: std.mem.Alignment, ret_addr: us
         return null;
     const result = scope.parent.vtable.alloc(scope.parent.ptr, len, alignment, ret_addr) orelse
         return null;
+    scope.trackAllocationAssumeCapacity(result[0..len], ret_addr, .none);
+    return result;
+}
+
+fn trackAllocationAssumeCapacity(scope: *AllocationScope, buf: []const u8, ret_addr: usize, extra: Extra) void {
     const trace = StoredTrace.capture(ret_addr);
-    scope.state.allocations.putAssumeCapacityNoClobber(result, .{
+    scope.state.allocations.putAssumeCapacityNoClobber(buf.ptr, .{
         .allocated_at = trace,
-        .len = len,
+        .len = buf.len,
+        .extra = extra,
     });
-    scope.state.total_memory_allocated += len;
-    return result;
+    scope.state.total_memory_allocated += buf.len;
 }
 
 fn free(ctx: *anyopaque, buf: []u8, alignment: std.mem.Alignment, ret_addr: usize) void {
     const scope: *AllocationScope = @ptrCast(@alignCast(ctx));
     scope.state.mutex.lock();
     defer scope.state.mutex.unlock();
-    var invalid = false;
+    const invalid = scope.trackFreeAssumeLocked(buf, ret_addr);
```

Comment:
would this return true for 0?
```
const buf = alloc(u8, 0);
defer free(buf);
```

Reply:
std.mem.Allocator special cases zero length allocations and frees to do nothing